File: nn_server/nn_spelling_input/nn_spelling_input.js
const nn_input = require("../nn_input/nn_input.js");
const spellChecker = require("simple-spellchecker");

const wordRegex = /\b[^\s]+\b/g;

// global in this file to prevent each class creating a new dictionary
const dictionaryUS = spellChecker.getDictionarySync("en-US"); // American
const dictionaryGB = spellChecker.getDictionarySync("en-GB"); // British

/**
 * @description a class which determines the fraction of english spelling errors.
 * @author Alistair Payn
 */
class nn_spelling_input extends nn_input {
    constructor() {
	super()
    }

  /**
   * @description determines the fraction of english spelling errors.
   * @author Alistair Payn
   */
  process(sentences) {
    let spellingErrorCount = 0;
    let wordCount = 0;
    sentences.forEach((sentence) => {
      let words = sentence.match(wordRegex);
      words.forEach((word) => {
        ++wordCount;
        spellingErrorCount +=
          !dictionaryUS.spellCheck(word) && !dictionaryGB.spellCheck(word);
      });
    });
    let result = 0.0;
    if (wordCount > 0) {
      result = spellingErrorCount / wordCount;
    }
    return result;
  }
}

module.exports = nn_spelling_input;

File: nn_server/nn_spelling_input/nn_spelling_input.py
import sys, os
dirname = os.path.dirname(__file__)
sys.path.append(os.path.join(dirname, '..', 'nn_input'))
from nn_input import NNInput
import nltk
from nltk.tokenize import word_tokenize
from spellchecker import SpellChecker
from autocorrect import Speller

class NNSpellingInput(NNInput):
    def __init__(self):
        super().__init__()
        self.__spellcheck = Speller('en')
        self.__autocorrect = SpellChecker()

    def process(self, sentences):
        totalErrors = 0
        totalWords = 0
        for s in sentences:
            tokens = word_tokenize(s.lower())
            for t in tokens:
                totalWords += 1                
                if t != self.__spellcheck(t) or t != self.__autocorrect.correction(t):
                    totalErrors += 1
        if totalWords != 0:
            return totalErrors / totalWords
        else:
            return 0.0
